aten/src/ATen/native/UpSample.h

```diff
@@ -0,0 +1,249 @@
+}
+
+static inline void upsample_3d_shape_check(
+    const Tensor& data,
+    int64_t type_check,
+    int64_t nbatch,
+    int64_t nchannels,
+    int64_t input_depth,
+    int64_t input_height,
+    int64_t input_width,
+    int64_t output_depth,
+    int64_t output_height,
+    int64_t output_width) {
+  AT_CHECK(
+      input_depth > 0 && input_height > 0 && input_width > 0 &&
+          output_depth > 0 && output_height > 0 && output_width > 0,
+      "Input and output sizes should be greater than 0, but got input (D: ",
+      input_depth,
+      ", H: ",
+      input_height,
+      ", W: ",
+      input_width,
+      ") output (D: ",
+      output_depth,
+      ", H: ",
+      output_height,
+      ", W: ",
+      output_width,
+      ")");
+
+  if (type_check == 0) {
+    AT_CHECK(
+        data.dim() == 5, "5D data tensor expected but got: ", data.dim(), "D");
+  } else if (type_check == 1) {
+    check_dim_size(data, 5, 0, nbatch);
+    check_dim_size(data, 5, 1, nchannels);
+    check_dim_size(data, 5, 2, output_depth);
+    check_dim_size(data, 5, 3, output_height);
+    check_dim_size(data, 5, 4, output_width);
+  }
+}
+
+template <typename scalar_t>
+static inline scalar_t linear_upsample_compute_scale(
+    int64_t input_size,
+    int64_t output_size,
+    bool align_corners) {
+  /* We view each pixel as an area, idx + 0.5 as its center index.
+   * Here is an example formula in 1D case.
+   * if align_corners: center of two corner pixel areas are preserved,
+   *     (0.5, 0.5) -> (0.5, 0.5),
+   *     (input_size - 0.5, 0.5) -> (output_size - 0.5)
+   *     scale = (input_size - 0.5 - 0.5) / (output_size - 0.5 - 0.5)
+   *     src_index + 0.5 - 0.5 = scale * (dst_index + 0.5 - 0.5)
+   * if not align_corners: the whole range is scaled accordingly
+   *     scale = input_size / output_size
+   *     src_idx + 0.5 = scale * (dst_index + 0.5)
+   */
+  if (output_size > 1) {
+    return align_corners
+        ? static_cast<scalar_t>(input_size - 1) / (output_size - 1)
+        : static_cast<scalar_t>(input_size) / output_size;
+  } else {
+    return scalar_t(0);
+  }
+}
+
+template <typename scalar_t>
+static inline scalar_t linear_upsample_compute_source_index(
+    scalar_t scale,
+    int64_t dst_index,
+    bool align_corners) {
+  if (align_corners) {
+    return scale * dst_index;
+  } else {
+    scalar_t src_idx = scale * (dst_index + 0.5) - 0.5;
+    return src_idx < 0 ? scalar_t(0) : src_idx;
+  }
+}
+
+static inline int64_t nearest_neighbor_compute_source_index(
+    const float scale,
+    int64_t dst_index,
+    int64_t input_size) {
+  const int64_t src_index =
+      std::min(static_cast<int64_t>(floorf(dst_index * scale)), input_size - 1);
+  return src_index;
+}
+
+template <typename scalar_t>
+static scalar_t upsample_get_value_bounded(
+    scalar_t* data,
+    int64_t width,
+    int64_t height,
+    int64_t x,
+    int64_t y) {
+  int64_t access_x = std::max(std::min(x, width - 1), static_cast<int64_t>(0));
```

Comment:
I'm a little surprised you need the static casts on 0 here.

Reply:
if I recall correctly here it raised an error in that time ... if you want I can try again with just 0.

Reply:
I do recall `std::max` being persnickety about the types of its overload. I'm happy to leave it be. Note that you can explicitly request a long long literal by writing `0LL`.

Reply:
awesome. I will use that! thanks!

Reply:
Yeah, I guess `0LL` doesn't work because `long long` is not the same thing as `int64_t`. The original way is fine, sorry about the goose chase.